src/java/native/libcephrgw_jni.cc

```diff
@@ -0,0 +1,997 @@
+		return(ret);
+	}
+
+	ret = rgw_open( rgw_fs, rgw_fh, flags, RGW_OPEN_FLAG_NONE );
+	ldout( cct, 10 ) << "jni: open: exit ret " << ret << dendl;
+
+	env->ReleaseStringUTFChars( j_path, c_path );
+	if ( ret < 0 )
+		handle_error( env, ret );
+
+	return( (jlong) rgw_fh);
+}
+
+
+JNIEXPORT jint JNICALL native_ceph_close( JNIEnv *env, jclass clz, jlong j_rgw_fs, jlong j_fd )
+{
+	struct rgw_fs		*rgw_fs = (struct rgw_fs *) j_rgw_fs;
+	int			ret;
+
+	ldout( cct, 10 ) << "jni: close: fd " << (long) j_fd << dendl;
+	ret = rgw_close( rgw_fs, (struct rgw_file_handle *)j_fd, RGW_CLOSE_FLAG_RELE );
+	ldout( cct, 10 ) << "jni: close: ret " << ret << dendl;
+
+	if ( ret )
+		handle_error( env, ret );
+
+	return(ret);
+}
+
+
+JNIEXPORT jlong JNICALL native_ceph_read( JNIEnv *env, jclass clz,
+    jlong j_rgw_fs, jlong j_fd, jlong j_offset, jbyteArray j_buf, jlong j_size )
+{
+	struct rgw_fs		*rgw_fs = (struct rgw_fs *) j_rgw_fs;
+	struct rgw_file_handle	*fh	= (struct rgw_file_handle *) j_fd;
+	jboolean		iscopy	= JNI_FALSE;
+	size_t			bytes_read;
+	jsize			buf_size;
+	jbyte			*c_buf;
+	long			ret;
+
+	CHECK_ARG_NULL( j_buf, "@buf is null", -1 );
+	CHECK_ARG_BOUNDS( j_size < 0, "@size is negative", -1 );
+
+	buf_size = env->GetArrayLength( j_buf );
+	CHECK_ARG_BOUNDS( j_size > buf_size, "@size > @buf.length", -1 );
+
+	c_buf = env->GetByteArrayElements( j_buf, &iscopy );
+	if ( !c_buf )
+	{
+		cephThrowInternal( env, "failed to pin memory" );
+		return(-1);
+	}
+
+	ldout( cct, 10 ) << "jni: read: fd " << (int) j_fd << " len " << (long) j_size <<
+	" offset " << (long) j_offset << dendl;
+
+	ret = rgw_read( rgw_fs, fh, (long) j_offset, (long) j_size, &bytes_read, c_buf, RGW_READ_FLAG_NONE );
+	ldout( cct, 10 ) << "jni: read: exit ret " << ret << "bytes_read " << bytes_read << dendl;
+	if ( ret < 0 )
+	{
+		handle_error( env, (int) ret );
+		bytes_read = 0;
+	}
+
+	env->ReleaseByteArrayElements( j_buf, c_buf, JNI_COMMIT );
+	if ( iscopy )
+	{
+		free( c_buf );
+	}
+
+	return( (jlong) bytes_read);
+}
+
+
+JNIEXPORT jlong JNICALL native_ceph_write( JNIEnv *env, jclass clz,
+    jlong j_rgw_fs, jlong j_fd, jlong j_offset, jbyteArray j_buf, jlong j_size )
+{
+	struct rgw_fs		*rgw_fs = (struct rgw_fs *) j_rgw_fs;
+	struct rgw_file_handle	*fh	= (struct rgw_file_handle *) j_fd;
+	size_t			bytes_written;
+	jsize			buf_size;
+	jbyte			*c_buf;
+	long			ret;
+
+	CHECK_ARG_NULL( j_buf, "@buf is null", -1 );
+	CHECK_ARG_BOUNDS( j_size < 0, "@size is negative", -1 );
+
+	buf_size = env->GetArrayLength( j_buf );
+	CHECK_ARG_BOUNDS( j_size > buf_size, "@size > @buf.length", -1 );
+
+	c_buf = env->GetByteArrayElements( j_buf, 0 );
+	if ( !c_buf )
+	{
+		cephThrowInternal( env, "failed to pin memory" );
+		return(-1);
+	}
+
+	ldout( cct, 10 ) << "jni: write: fd " << (int) j_fd << " len " << (long) j_size <<
+	" offset " << (long) j_offset << dendl;
+
+	ret = rgw_write( rgw_fs, fh, (long) j_offset, (long) j_size, &bytes_written, c_buf, RGW_WRITE_FLAG_NONE );
```

Comment:
should be ok, but would be ideal to use platform size types, or perhaps stdint.h types?

Reply:
OK, we'll use uint64_t instead of long.